pkg/manifests/external_dns.go

```diff
@@ -54,41 +86,45 @@ func (p Provider) String() string {
 	}
 }
 
-func (p Provider) ResourceName() string {
-	switch p {
-	case PublicProvider:
-		return externalDnsResourceName
-	case PrivateProvider:
-		return externalDnsResourceName + "-private"
-	default:
-		return ""
+// ExternalDnsConfig defines configuration options for required resources for external dns
+type ExternalDnsConfig struct {
+	TenantId, Subscription, ResourceGroup,
+	ClientId, ServiceAccountName, Namespace,
+	CRDName string
+	IdentityType       IdentityType
+	ResourceTypes      []ResourceType
+	Provider           Provider
+	DnsZoneResourceIDs []string
+}
+
+func (e *ExternalDnsConfig) resourceName() string {
+	if e.CRDName == "" {
+		switch e.Provider {
+		case PublicProvider:
+			return externalDnsResourceName
+		case PrivateProvider:
+			return externalDnsResourceName + "-private"
+		}
 	}
+	return e.CRDName
+
 }
 
-func (p Provider) Labels() map[string]string {
+func ExternalDNSLabels(e *ExternalDnsConfig) map[string]string {
 	labels := map[string]string{
-		k8sNameKey: p.ResourceName(),
+		k8sNameKey: e.resourceName(),
 	}
 	return labels
 }
 
-// ExternalDnsConfig defines configuration options for required resources for external dns
-type ExternalDnsConfig struct {
-	TenantId, Subscription, ResourceGroup string
-	Provider                              Provider
-	DnsZoneResourceIDs                    []string
-}
-
 // ExternalDnsResources returns Kubernetes objects required for external dns
 func ExternalDnsResources(conf *config.Config, externalDnsConfigs []*ExternalDnsConfig) []client.Object {
 	var objs []client.Object
-
-	// Can safely assume the namespace exists if using kube-system
-	if conf.NS != "kube-system" {
-		objs = append(objs, Namespace(conf))
-	}
-
 	for _, dnsConfig := range externalDnsConfigs {
+		// Can safely assume the namespace exists if using kube-system
+		if dnsConfig.Namespace != "" && dnsConfig.Namespace != "kube-system" {
```

Comment:
where do we default to a namespace in dnsConfig?

we the default should be conf.NS (that's the old behavior)

Reply:
we should really control the constructor for the externalDns configs within this package to prevent misuse

Reply:
I agree with the use of the constructor - i'll make the type private and add a public constructor, so that there's no need to default on a namespace